lib/logger_json/jason_safe_formatter.ex

```diff
@@ -0,0 +1,79 @@
+defmodule LoggerJSON.JasonSafeFormatter do
+  @moduledoc """
+  Utilities for converting metadata into datastructures that can be safely passed to Jason.encode!/1
+  """
+
+  @doc """
+  Produces metadata that is "safe" for calling Jason.encode!/1 on without errors.
+  This means that unexpected Logger metadata won't cause logging crashes.
+  Current formatting is...
+  - Maps: as is
+  - Printable binaries: as is
+  - Numbers: as is
+  - Structs that don't implement Jason.Encoder: converted to maps
+  - Tuples: converted to lists
+  - Keyword lists: converted to Maps
+  - everything else: inspected
+  """
+  @spec format(any()) :: any()
+  def format(%Jason.Fragment{} = data) do
+    data
+  end
+
+  def format(nil), do: nil
+  def format(true), do: true
+  def format(false), do: false
+  def format(data) when is_atom(data), do: data
+
+  def format(%mod{} = data) do
+    if jason_implemented?(mod) do
+      data
+    else
+      data
+      |> Map.from_struct()
+      |> format()
+    end
+  end
+
+  def format(%{} = data) do
+    for {key, value} <- data, into: %{}, do: {key, format(value)}
+  end
+
+  def format([{key, _} | _] = data) when is_atom(key) do
+    Enum.into(data, %{}, fn
+      {key, value} -> {key, format(value)}
+    end)
+  rescue
+    _ -> for(d <- data, do: format(d))
+  end
+
+  def format({key, data}) when is_binary(key) or is_atom(key), do: %{key => format(data)}
+
+  def format(data) when is_tuple(data), do: Tuple.to_list(data)
+
+  def format(data) when is_number(data), do: data
+
+  def format(data) when is_binary(data) do
+    if String.valid?(data) && String.printable?(data) do
+      data
+    else
+      inspect(data)
+    end
+  end
+
+  def format(data) when is_list(data), do: for(d <- data, do: format(d))
+
+  def format(data) do
+    inspect(data, pretty: true, width: 80)
+  end
+
+  def jason_implemented?(mod) do
+    try do
```

Comment:
Can we replace `try` block with `Jason.Encoder.impl_for/1`? Try blocks are very expensive and we need to make sure that logger formatting is efficient

Reply:
Oh nice, TIL. I had no idea about `impl_for/1`. Where is that even documented?

Anyway, I changed over to that and it seems to work, tests are all passing in this repo and in our App test suite and avoiding the try/catch will be a nice win.

I will look to try this in our App in production next week but it looks good so thanks for the pointer.

Reply:
It's documented in `Protocol` module doc, section "Reflection". However, I agree that if you never saw that doc you will never find out that it exists :(. I think Elixir team would be happy to accept PR that improves docs in that regard :)